gui/halt_list_frame.cc

```diff
@@ -110,6 +113,49 @@ bool halt_list_frame_t::compare_halts(halthandle_t const halt1, halthandle_t con
 		case nach_typ: // sort by station type
 			order = halt1->get_station_type() - halt2->get_station_type();
 			break;
+		case nach_transfer: // sort by no of transfers in previous month
+		{
+			int id1 = halt1->get_finance_history( 1, HALT_ARRIVED) == 0 ||
+				halt1->get_finance_history( 1, HALT_DEPARTED) == 0 ? 0 : 1;
+			auto halt1_transfers = halt1->get_finance_history(id1, HALT_ARRIVED)
+				< halt1->get_finance_history(id1, HALT_DEPARTED) ?
+				halt1->get_finance_history(id1, HALT_ARRIVED) :
+				halt1->get_finance_history(id1, HALT_DEPARTED);
+			int id2 = halt2->get_finance_history( 1, HALT_ARRIVED) == 0 ||
+				halt2->get_finance_history( 1, HALT_DEPARTED) == 0 ? 0 : 1;
+			auto halt2_transfers = halt2->get_finance_history(id2, HALT_ARRIVED)
+				< halt2->get_finance_history(id2, HALT_DEPARTED) ?
+				halt2->get_finance_history(id2, HALT_ARRIVED) :
+				halt2->get_finance_history(id2, HALT_DEPARTED);
+			order = halt1_transfers > halt2_transfers ? 1 :
+				halt1_transfers < halt2_transfers ? -1 : 0;
+			break;
+		}
+		case nach_wartend_percent: // sort by waiting goods percentage
+		{
+			sint64 sum1 = 0;
+			sint64 sum2 = 0;
+			for(unsigned int i=0; i<goods_manager_t::get_count(); i++) {
+				const goods_desc_t *wtyp = goods_manager_t::get_info(i);
+				if(halt1->gibt_ab(wtyp)) {
+					sum1 += halt1->get_capacity( i>2?2:i );
+				}
+				if(halt2->gibt_ab(wtyp)) {
+					sum2 += halt2->get_capacity( i>2?2:i );
+				}
+			}
+
+			if (sum1 == 0 && sum2 != 0) {
+    			order = -1; // Halt2 has non-zero sum, so it comes first
+			} else if (sum1 != 0 && sum2 == 0) {
+				order = 1; // Halt1 has non-zero sum, so it comes first
+			} else if (sum1 != 0 && sum2 != 0) {
+				double per1 = (double) halt1->get_finance_history( 0, HALT_WAITING )/sum1 * 100;
+				double per2 = (double) halt2->get_finance_history( 0, HALT_WAITING )/sum2 * 100;
+				order = per1 > per2 ? 1 : (per1 < per2 ? -1 : 0);
```

Comment:
(nit) To avoid the rounding error, always prefer the integer calculation. In this case, the order can be determined by calculating `halt1->get_finance_history( 0, HALT_WAITING ) * sum2` and `halt2->get_finance_history( 0, HALT_WAITING ) * sum1`.